File: org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/ui/popup/PopupManager.java
/*
 * Copyright (c) 2010, 2023 BSI Business Systems Integration AG
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.scout.rt.client.ui.popup;

import java.util.Set;

import org.eclipse.scout.rt.platform.Bean;
import org.eclipse.scout.rt.platform.reflect.AbstractPropertyObserver;
import org.eclipse.scout.rt.platform.util.CollectionUtility;

/**
 * @since 9.0
 */
@Bean
public class PopupManager extends AbstractPropertyObserver {

  public static final String PROP_POPUPS = "popups";

  public PopupManager() {
    propertySupport.setPropertySet(PROP_POPUPS, CollectionUtility.emptyHashSet());
  }

  public Set<IPopup> getPopups() {
    return CollectionUtility.hashSet(getPopupsInternal());
  }

  protected Set<IPopup> getPopupsInternal() {
    return propertySupport.getPropertySet(PROP_POPUPS);
  }

  @SuppressWarnings("unchecked")
  public <T extends IPopup> T getPopupByClass(Class<T> popupClass) {
    return (T) getPopupsInternal().stream().filter(popupClass::isInstance).findFirst().orElse(null);
  }

  protected void setPopupsInternal(Set<? extends IPopup> popups) {
    propertySupport.setPropertySet(PROP_POPUPS, popups);
  }

  public void open(IPopup popup) {
    Set<IPopup> popups = getPopups();
    popups.add(popup);
    setPopupsInternal(popups);
  }

  public void close(IPopup popup) {
    Set<? extends IPopup> popups = getPopups();
    popups.remove(popup);
    setPopupsInternal(popups);
  }
}

File: eclipse-scout-core/src/popup/PopupManager.js
/*
 * Copyright (c) 2018 BSI Business Systems Integration AG.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     BSI Business Systems Integration AG - initial API and implementation
 */
import {arrays, Widget} from '../index';

export default class PopupManager extends Widget {

  constructor() {
    super();
    this.popups = [];
    this._addWidgetProperties(['popups']);
    this._addPreserveOnPropertyChangeProperties(['popups']);
  }

  _init(model) {
    super._init(model);
    this.session.layoutValidator.schedulePostValidateFunction(() => {
      this._openPopups(this.popups);
    });
  }

  setPopups(popups) {
    this.setProperty('popups', popups);
  }

  _setPopups(popups) {
    this._openPopups(arrays.diff(popups, this.popups));
    this._destroyPopups(arrays.diff(this.popups, popups));
    this._setProperty('popups', popups);
    // re-parent popups, since PopupManager is not a real widget but only used to sync data
    this.popups.forEach(popup => {
      if (popup.parent instanceof scout.PopupManager) {
        popup.setParent(this.session.desktop);
      }
    });
  }

  _openPopups(popups) {
    popups.forEach(popup => {
      popup.open(); // Let the popup itself determine the entry point (important for popup windows)
    });
  }

  _destroyPopups(popups) {
    popups.forEach(popup => {
      popup.destroy();
    });
  }
}
